src/store/io.rs

```diff
@@ -0,0 +1,166 @@
+}
+
+pub struct CompleteIo {
+    command: IoCommand,
+    result: std::io::Result<()>,
+}
+
+/// Create an I/O worker managing an io_uring and sending responses back via channels to a number
+/// of handles.
+pub fn start_io_worker(
+    store: Arc<Store>,
+    num_handles: usize,
+) -> (Sender<IoCommand>, Vec<Receiver<CompleteIo>>) {
+    // main bound is from the pending slab.
+    let (command_tx, command_rx) = crossbeam_channel::bounded(1);
+    let (handle_txs, handle_rxs) = (0..num_handles)
+        .map(|_| crossbeam_channel::bounded(32))
+        .unzip();
+    std::thread::spawn(move || run_worker(store, command_rx, handle_txs));
+    (command_tx, handle_rxs)
+}
+
+// hack: this is not exposed from the io_uring or libc libraries.
+const IORING_ENTER_GETEVENTS: u32 = 1;
+
+fn run_worker(
+    store: Arc<Store>,
+    command_rx: Receiver<IoCommand>,
+    handle_tx: Vec<Sender<CompleteIo>>,
+) {
+    let mut pending: Slab<IoCommand> = Slab::with_capacity(SLAB_CAPACITY);
+
+    let mut ring = IoUring::<squeue::Entry, cqueue::Entry>::builder()
+        .setup_single_issuer()
+        .setup_sqpoll(100)
+        // .setup_iopoll
+        .build(RING_CAPACITY)
+        .expect("Error building io_uring");
+
+    loop {
+        // note: dropping the queues at the end of loop iteration performs `sync` implicitly
+        let (submitter, mut submit_queue, mut complete_queue) = ring.split();
+
+        // 1. process completions.
+        if !pending.is_empty() {
+            // block on next completion if at capacity.
+            if pending.len() == SLAB_CAPACITY && complete_queue.is_empty() {
+                // TODO: handle error
+                unsafe {
+                    submitter
+                        .enter::<libc::sigset_t>(
+                            0, // submit
+                            1, // complete
+                            IORING_ENTER_GETEVENTS,
+                            None,
+                        )
+                        .unwrap();
+                }
+                complete_queue.sync();
+            }
+
+            for completion_event in complete_queue {
+                let command = pending.remove(completion_event.user_data() as usize);
+                let handle_idx = command.handle;
+                let result = if completion_event.result() != 0 {
+                    Err(std::io::Error::from_raw_os_error(completion_event.result()))
+                } else {
+                    Ok(())
+                };
+                let complete = CompleteIo { command, result };
+                if let Err(_) = handle_tx[handle_idx].send(complete) {
+                    // TODO: handle?
+                    break;
+                }
+            }
+        }
+
+        // 2. accept new I/O requests when slab has space & submission queue is non-empty.
+        let mut to_submit = false;
+        submit_queue.sync();
+        while pending.len() < SLAB_CAPACITY && !submit_queue.is_full() {
```

Comment:
At this point it woulda been better to rename `SLAB_CAPACITY` to something like `MAX_INFLIGHT`.

Reply:
yeah, agreed